runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+		/// a particular collator has already been chosen for the next block, and no other collator
+		/// may provide the block. In this case we allow the possibility of the combination being
+		/// retried in a later block, expressed by `Retriable`.
+		///
+		/// Ordered as Parachains ++ Selected_Parathreads.
+		Active: Vec<(ParaId, Option<(CollatorId, Retriable)>)>;
+
+		/// The next unused ParaId value. Start this high in order to keep low numbers for
+		/// system-level chains.
+		NextFreeId: ParaId = LOWEST_USER_ID;
+
+		/// Pending swap operations.
+		PendingSwap: map ParaId => Option<ParaId>;
+
+		/// Map of all registered parathreads/chains.
+		Paras get(paras): map ParaId => Option<ParaInfo>;
+
+		/// The current queue for parathreads that should be retried.
+		RetryQueue get(retry_queue): [Vec<(ParaId, CollatorId)>; MAX_RETRIES as usize];
+
+		/// Users who have paid a parathread's deposit
+		Debtors: map ParaId => T::AccountId;
+	}
+	add_extra_genesis {
+		config(parachains): Vec<(ParaId, Vec<u8>, Vec<u8>)>;
+		config(_phdata): PhantomData<T>;
+		build(build::<T>);
+	}
+}
+
+#[cfg(feature = "std")]
+fn build<T: Trait>(config: &GenesisConfig<T>) {
+	let mut p = config.parachains.clone();
+	p.sort_unstable_by_key(|&(ref id, _, _)| *id);
+	p.dedup_by_key(|&mut (ref id, _, _)| *id);
+
+	let only_ids: Vec<ParaId> = p.iter().map(|&(ref id, _, _)| id).cloned().collect();
+
+	Parachains::put(&only_ids);
+
+	for (id, code, genesis) in p {
+		Paras::insert(id, &primitives::parachain::PARACHAIN_INFO);
+		// no ingress -- a chain cannot be routed to until it is live.
+		<parachains::Code>::insert(&id, &code);
+		<parachains::Heads>::insert(&id, &genesis);
+		<parachains::Watermarks<T>>::insert(&id, T::BlockNumber::zero());
+		// Save initial parachains in registrar
+		Paras::insert(id, ParaInfo { scheduling: Scheduling::Always })
+	}
+}
+
+pub fn swap_ordered_existence<T: PartialOrd + Ord + Copy>(ids: &mut Vec<T>, one: T, other: T) {
+	let maybe_one_pos = ids.binary_search(&one);
+	let maybe_other_pos = ids.binary_search(&other);
+	match (maybe_one_pos, maybe_other_pos) {
+		(Ok(one_pos), Err(_)) => ids[one_pos] = other,
+		(Err(_), Ok(other_pos)) => ids[other_pos] = one,
+		_ => return,
+	};
+	ids.sort();
+}
+
+decl_module! {
+	/// Parachains module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		fn deposit_event() = default;
+
+		/// Register a parachain with given code.
+		/// Fails if given ID is already used.
+		#[weight = SimpleDispatchInfo::FixedOperational(5_000_000)]
+		pub fn register_para(origin,
+			#[compact] id: ParaId,
+			code: Vec<u8>,
+			initial_head_data: Vec<u8>,
+			info: ParaInfo
+		) -> Result {
+			ensure_root(origin)?;
+			<Self as Registrar<T::AccountId>>::
+				register_para(id, info, code, initial_head_data)
+		}
+
+		/// Deregister a parachain with given id
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000)]
```

Comment:
this is actually an `O(n)` operation. might be optimizable with a `double_map`

Reply:
well, insofar as it's a direct dispatch, it's a root call, and there are many other instances of O(N) or worse operation on root dispatches; we assume that the governance can live with the costs if it's dispatching them.

it is, however, used by `deregister_parathread`, which is more problematic. we probably just want to either cancel all queued messages, limit the queue size or require the queue be processed in full before deregistration. i'm not sure which is best, but i'm also not convinced that incoming message queue management (which needs to be addressed even for regular parachains that for whatever reason have sporadic dispatching) is really in the scope of this PR.